core/src/main/java/io/parsingdata/metal/token/Tie.java

```diff
@@ -59,24 +60,24 @@ public Tie(final String name, final Token token, final ValueExpression dataExpre
     }
 
     @Override
-    protected Optional<Environment> parseImpl(final String scope, final Environment environment, final Encoding encoding) {
-        final ImmutableList<Optional<Value>> dataResult = dataExpression.eval(environment.order, encoding);
+    protected Optional<ParseState> parseImpl(final String scope, final ParseState parseState, final Callbacks callbacks, final Encoding encoding) {
+        final ImmutableList<Optional<Value>> dataResult = dataExpression.eval(parseState, encoding);
         if (dataResult.isEmpty()) {
             return failure();
         }
-        return iterate(scope, dataResult, 0, environment, environment.addBranch(this), encoding).computeResult();
+        return iterate(scope, dataResult, 0, parseState, parseState.addBranch(this), callbacks, encoding).computeResult();
     }
 
-    private Trampoline<Optional<Environment>> iterate(final String scope, final ImmutableList<Optional<Value>> values, final int index, final Environment returnEnvironment, final Environment environment, final Encoding encoding) {
+    private Trampoline<Optional<ParseState>> iterate(final String scope, final ImmutableList<Optional<Value>> values, final int index, final ParseState returnParseState, final ParseState parseState, final Callbacks callbacks, final Encoding encoding) {
         if (values.isEmpty()) {
-            return complete(() -> success(new Environment(environment.closeBranch().order, returnEnvironment.source, returnEnvironment.offset, returnEnvironment.callbacks)));
+            return complete(() -> success(new ParseState(parseState.closeBranch().order, returnParseState.source, returnParseState.offset)));
```

Comment:
Not using a factory method here?

Reply:
I suppose that would have to be a method comparable to `seek()` that replaces the `ParseGraph` on `returnEnvironment`. It could be done, but this seems to be a rather isolated use.